### ISSUE TITLE
don't warn if ul tag without li as specified for html5

### ISSUE THREAD

**Issue by rtack** [htacg/tidy-html5#396]
https://www.w3.org/TR/html-markup/ul.html
says:
Permitted content:
Zero or more li elements

Following code get's a warning displayed even though it's valid:

```
<ul class="list-group">
    <a href="#submenu" class="list-group-item active" data-toggle="collapse" data-target="">someText</a>
...
```

line 2 column 1 - Warning: missing `<li>`

Edit: Added backcheck to element

**Comment by geoffmcl** [comment]
@rtack thanks for the report...

I got quite a **surprise** when running your simple sample through a current tidy version 5.1.51, with the config `--show-body-only yes --show-info no`... 

It **converted** the `<ul class="list-group">` to a `<div...>`???

Output:

````
line 3 column 5 - Warning: missing <li>
line 2 column 1 - Warning: missing </ul>
Tidy found 2 warnings and 0 errors!

Info: Document content looks like HTML5
<div style="margin-left: 2em" class="list-group"><a href="#submenu"
class="list-group-item active" data-toggle="collapse" data-target=
"">someText</a></div>
````

So **yes** there is the warning about the missing `<li>`, and assume this would only apply to html5, and **not** to legacy html4-- documents? That is one thing to check and verify... Maybe the `Zero or more li elements` only applies to html5...

Then it also reports missing `</ul>`, but there is now no `<ul>`! It removed it... so why this report... maybe a minor niggle...

Really need help on understanding what is happening here! Under what circumstance will current tidy do this conversion? And why?

So at this point can do nothing until I understand more... comments, ideas, help **very** welcome... thanks...

**Comment by rtack** [comment]
yes I believe that in html4 it was 1 or more li.
in html5 it's 0 or more.

I plan looking into what's going sometime if nobody beats me to it.

converting ul to div is possibly (most likely?) a different issue.

**Comment by rtack** [comment]
It seems to be the case that with html4 at least one li element was required and with html5 it's 0 or more.

html4 spec for ul tag: https://www.w3.org/TR/html401/struct/lists.html#h-10.2
```
<!ELEMENT UL - - (LI)+                 -- unordered list -->
```
html5 spec for ul tag: https://www.w3.org/TR/html-markup/spec.html#ul
```
Permitted content:
Zero or more li elements
```


with tidy version 5.1.51
```
<!DOCTYPE html>
<html>
<head>
  <title>any</title>
</head>
<body>
  <ul>
    <a href="#submenu"></a>
  </ul>
</body>
</html>
```
and no config will result in the following

```
line 10 column 5 - Warning: missing <li>
Info: Document content looks like HTML5
Tidy found 1 warning and 0 errors!

<!DOCTYPE html>
<html>
<head>
<meta name="generator" content=
"HTML Tidy for HTML5 for Linux version 5.1.51">
<title>any</title>
</head>
<body>
<div style="margin-left: 2em"><a href="#submenu"></a></div>
</body>
</html>
```

**Comment by geoffmcl** [comment]
@rtack, wow, like the previous issue #395, thank you for the feedback, further testing, version information, and the **time consuming research**...

As you have no doubt understood, tidy has basically two modes. It starts, defaults to html5++ mode, but if it finds a legacy doctype in the data, or is given a user `--doctype <string>`, can revert to html4-- mode...

At present there is no simple `flag` to show this, but it has been in my mind for quite a while that there should be one... like a simple internal `TY_(IsHTML5Mode)(doc)`, return `yes`, `no`...

This could be a `Bool isHTML5` value added to the opaque `_Lexer` structure, which is re-initialized for each document, and could be set by `TY_(ResetTags)`, and un-set by `TY_(AdjustTags)`... need to check the code exactly...

Some places in tidy use `TY_(HTMLVersion)(doc)`, and/or `TY_(ApparentVersion)(doc)`, **but** this should only be used after **all** the document parsing is **completed**... all the constraining values taken into account... not **during** the document parsing...

Then this new `TY_(IsHTML5Mode)(doc)` could influence whether this no `<li>` warning is issued... simple...

**BUT**, this is inexplicably mixed up with why does tidy choose to **convert** the `<ul>` to a `<div>`! 

To me this **must** be understood... This conversion might be a separate issue, but may influence what we do about this missing `<li>` warning, if it is no longer a `<ul>`... aside from the simple suggested mode check...

Also thank you for the simple `input` example... as stated this always helps... And personally, without further explanations, I can see no good reason to convert what looks like perfectly valid html5 `<ul> ... </ul>` input to be converted to a `<div>`!!!

One tidy niggle on the input title... why not make this like `Issue #396`... but not important...

So yes, let's add a `TY_(IsHTML5Mode)(doc)` service, and suppress this missing `<li>` warning if still html5++ mode... that would be one thing **fixed**... and could help in other issues like it..

As always, patch or PR **most** welcome...

But me, I still seek more understanding of the current **conversion** reasoning...

**Comment by geoffmcl** [comment]
@rtack, have carefully looked into the **conversion** reasoning... and maybe it is all ok...

First, tidy has been doing this for a **long** time! As far back as vers 4th August 2000! Practically the first release of tidy...

It is done **after** the document parsing, in `tidyCleanAndRepair(tdoc);`, specifically it begins in `TY_(List2BQ)()`.

As the name implies, this renames the `<ul>` to a `<blockquote>`, under certain conditions. Those conditions include that it only has **one** child, and that child is **implicit**. That is added by tidy. The code comment associated with this service is -

````
/*
 Some people use dir or ul without an li
 to indent the content. The pattern to
 look for is a list with a single implicit
 li. This is recursively replaced by an
 implicit blockquote.
*/
````

In this case, at that moment, the tidy node tree contains -

````
  StartTag body
   StartTag ul
    StartTag li implicit  style="list-style: none; display: inline"
     StartTag a   href="#submenu"
````

So one can see that tidy has already set this up by adding an implicit `<li style="list-style: none; display: inline">` around the anchor...

In `List2BQ`, the `<li>` is removed, just leaving the anchor contents, and the `<ul>` is renamed to `<blockquote>`, and marked as `implicit`.

The next transformation occurs in `TY_(BQ2Div)()`, again under certain conditions, namely that it is an implicit `<blockquote>`, which will now be renamed to a `<div>`, and the style `margin-left: 2em` added. The comment associated with this service is -

````
/*
 Replace implicit blockquote by div with an indent
 taking care to reduce nested blockquotes to a single
 div with the indent set to match the nesting depth
*/
````

And the change is complete... Now the tidy node tree contains -

````
  StartTag body  
   StartTag div implicit  style="margin-left: 2em"
    StartTag a   href="#submenu"
````

So we can see this would not have happened if tidy had **not** added that implicit `<li>`! So fixing this for html5 mode - that is **not** adding an implicit `<li>` during parsing - will also remove this transformation of `<ul>` to `<blockquote>` to `<div>`... my understanding of this is now complete...

So backing up into the parsing, at the point where an implicit `<li>` is added, and the warning issued...

The problem starts in [parser.c:2517](https://github.com/htacg/tidy-html5/blob/master/src/parser.c#L2517). The next token in the stream is **not** an `<li>` - ie `if ( !nodeIsLI(node) )`. The token is stored, and an implicit node added...

If we could check the mode, like `if ( !nodeIsLI(node) && !TY_(IsHTML5Mode)(doc) )` then things would be good... this next token would be added as a child of the `<ul>`, what ever it is...

I will work on this, if you or others do not beat me to it... thanks...

**Comment by geoffmcl** [comment]
It turns out there are some good results in fixing this **bug**... See Issue #310 

Patch to build **5.1.52-test1**

````diff
diff --git a/src/parser.c b/src/parser.c
index 0447fe8..ae773aa 100644
--- a/src/parser.c
+++ b/src/parser.c
@@ -328,7 +328,6 @@ static Bool CanPrune( TidyDocImpl* doc, Node *element )
     if (nodeIsDD(element))
         return no;
 
-
     return yes;
 }
 
@@ -2514,7 +2513,14 @@ void TY_(ParseList)(TidyDocImpl* doc, Node *list, GetTokenMode ARG_UNUSED(mode))
             continue;
         }
 
-        if ( !nodeIsLI(node) )
+        if ( nodeIsLI(node) || TY_(IsHTML5Mode)(doc))
+        {
+            /* node is <LI> 
+               Issue #396 - A <ul> can have Zero or more li elements
+             */
+            TY_(InsertNodeAtEnd)(list,node);
+        }
+        else
         {
             TY_(UngetToken)( doc );
 
@@ -2567,9 +2573,6 @@ void TY_(ParseList)(TidyDocImpl* doc, Node *list, GetTokenMode ARG_UNUSED(mode))
                 TY_(InsertNodeAtEnd)(list,node);
             }
         }
-        else
-            /* node is <LI> */
-            TY_(InsertNodeAtEnd)(list,node);
 
         ParseTag( doc, node, IgnoreWhitespace);
     }
diff --git a/src/tags.c b/src/tags.c
index ed3436c..9e77fcb 100644
--- a/src/tags.c
+++ b/src/tags.c
@@ -780,6 +780,12 @@ void TY_(AdjustTags)( TidyDocImpl *doc )
     }
 }
 
+Bool TY_(IsHTML5Mode)( TidyDocImpl *doc )
+{
+    return doc->HTML5Mode;
+}
+
+
 /*\
  * Issue #285
  * Reset the table to default HTML5 mode.
@@ -809,6 +815,7 @@ void TY_(ResetTags)( TidyDocImpl *doc )
 #if ELEMENT_HASH_LOOKUP
     tagsEmptyHash( doc, tags ); /* not sure this is really required, but to be sure */
 #endif
+    doc->HTML5Mode = yes;   /* set HTML5 mode */
 }
 
 void TY_(FreeTags)( TidyDocImpl* doc )
@@ -823,6 +830,8 @@ void TY_(FreeTags)( TidyDocImpl* doc )
 
     /* get rid of dangling tag references */
     TidyClearMemory( tags, sizeof(TidyTagImpl) );
+
+    doc->HTML5Mode = no;    /* reset html5 mode == legacy html4 mode */
 }
 
 
diff --git a/src/tags.h b/src/tags.h
index 2412717..de75809 100644
--- a/src/tags.h
+++ b/src/tags.h
@@ -88,6 +88,7 @@ void TY_(InitTags)( TidyDocImpl* doc );
 void TY_(FreeTags)( TidyDocImpl* doc );
 void TY_(AdjustTags)( TidyDocImpl *doc ); /* if NOT HTML5 DOCTYPE, fall back to HTML4 legacy mode */
 void TY_(ResetTags)( TidyDocImpl *doc ); /* set table to HTML5 mode */
+Bool TY_(IsHTML5Mode)( TidyDocImpl *doc );
 
 /* Parser methods for tags */
 
diff --git a/src/tidy-int.h b/src/tidy-int.h
index bf1ccd3..e76cb1d 100755
--- a/src/tidy-int.h
+++ b/src/tidy-int.h
@@ -76,6 +76,8 @@ struct _TidyDocImpl
     uint                badChars;    /* for bad char encodings */
     uint                badForm;     /* bit field, for badly placed form tags, or other format errors */
 
+    Bool                HTML5Mode;  /* current mode is html5 */
+
     /* Memory allocator */
     TidyAllocator*      allocator;
 
diff --git a/version.txt b/version.txt
index 4901f78..5ca3a38 100644
--- a/version.txt
+++ b/version.txt
@@ -1,2 +1,2 @@
-5.1.52
+5.1.52-test1
 2016.04.04

````

Will try to create an `issue-396` branch soon, unless you beat me to it...

**Comment by rtack** [comment]
I'll create an issue-396 branch for testing #396 and #310

**Comment by rtack** [comment]
I've created a new branch on my fork. https://github.com/rtack/tidy-html5/tree/issue-396
do I need to do a normal pull request? or will you create a new branch from my branch?

**Comment by balthisar** [comment]
@rtack, just start a normal PR.

**Comment by geoffmcl** [comment]
@rtack, as @balthisar says, just add a normal PR, and I will deal with it... looks great in a quick review... 

Can't wait to tests when it is in htacg, as an `issue-396` branch... probably version **5.3.0-test2**... thanks...

**Comment by geoffmcl** [comment]
Now merged to `master`, and 5.3.1... thanks..

If you get a chance to re-test this issue, and if ok, close this... thanks..
